core/src/transport.rs

```diff
@@ -115,12 +108,20 @@ pub trait Transport {
     /// obtained from [dialing](Transport::dial).
     type Dial: Future<Output = Result<Self::Output, Self::Error>>;
 
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<TransportEvent<Self>>
+    where
+        Self: Sized;
+
     /// Listens on the given [`Multiaddr`], producing a stream of pending, inbound connections
-    /// and addresses this transport is listening on (cf. [`ListenerEvent`]).
+    /// and addresses this transport is listening on (cf. [`TransportEvent`]).
     ///
     /// Returning an error from the stream is considered fatal. The listener can also report
-    /// non-fatal errors by producing a [`ListenerEvent::Error`].
-    fn listen_on(&mut self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>>
+    /// non-fatal errors by producing a [`TransportEvent::Error`].
+    fn listen_on(
+        &mut self,
+        id: ListenerId,
```

Comment:
`Transport`s how have to be aware of `ListenerId`s since they need to include in `TransportEvent`s the information which listener produced the event. This is not ideal for users that use `core/transport` directly outside of a swarm, but I don't think there's any other way to solve this.

Reply:
> This is not ideal for users that use `core/transport` directly outside of a swarm, but I don't think there's any other way to solve this.

I think it is perfectly fine. IMO any serious user of `Transport` should care about unambiguously mapping attempts of listening on an address to the outcome.

Reply:
Unresolving here to continue the discussion whether the `ListenerId` as an argument is needed here.

If I understand correctly, `ListenerId` is passed as an argument to prevent to `Transport` implementations to have clashing IDs, correct?

Do we depend on the monotonicity of the ID anywhere? Asked differently, could the `ListenerId` as well be a random e.g. `u64`? With this approach, there is no need to pass the `ListenerId` as an argument, but instead can be generated by each `Transport` implementations without the need for synchronization.

Reply:
> could the `ListenerId` as well be a random e.g. `u64`

I would be in favor of making it an opaque type. We should probably remove the `PartialOrd` and `Ord` derives if we generate them randomly.

Reply:
With 2226092 I followed @thomaseizinger idea (https://github.com/libp2p/rust-libp2p/pull/2652#discussion_r874461833) to namespace `ListenerId`s to avoid clashing. The advantage of this implementation is that we can keep the monotonicity that I think all other Ids in libp2p have. Drawback is that it's a bit more complicated when we wrap inner transports. I have implemented it now that transports in `/transports` that wrap an inner transport with additional logic (like `WsConfig` and `DnsConfig`) change the TypeId of ListenerIds from inner transports, but wrapper transport in `core/transport` just pass it on. For the former case it is needed so user can compose a transport multiple times with different wrappers (like our development_transport does it with dns-tcp and ws-dns-tcp), for the latter I don't think the TypeId should be changed since the "actual" transport is still the same.
Also, wrapper transports that map the ListenerId need to also map it for all TransportEvent, but unfortunately this is not enforced on a type level which make is rather error-prone.

Alternative of making it a random id would also work and safe us a bit of complexity, but I am also content with the current solution.

Don't have a strong opinion here, I am fine with either. Wdyt @thomaseizinger @mxinden.

Reply:
> For the former case it is needed so user can compose a transport multiple times with different wrappers (like our development_transport does it with dns-tcp and ws-dns-tcp)

We never clone "leaf" transports like Tcp right? And only these transports generate `ListenerId`s. As long as we don't clone such a leaf transport, shouldn't the generated `ListenerId`s still be unique without mapping the type?

An alternative solution would be to offer a way of combining `ListenerId`s and even wrappers generate their own `ListenerId`s but combine them with the ones they get returned from inner transports. All of this seems very complicated though so perhaps choosing them randomly is better :)

Reply:
> We never clone "leaf" transports like Tcp right? And only these transports generate ListenerIds. As long as we don't clone such a leaf transport, shouldn't the generated ListenerIds still be unique without mapping the type?

Yes we clone "leaf" transports like tcp. For example in `development_transport`, we create two tcp transports wrapped in a dns transport, and then wrap one in a  WsConfig and combine them in a `OrTransport`. If we would only bubble up the ids generated by the tcp transport, both "or"-ed transports would generate the same Ids:
https://github.com/libp2p/rust-libp2p/blob/4aa84bf7cd78c8553294f4f352fce9e20b6fbd4e/src/lib.rs#L203-L208

> An alternative solution would be to offer a way of combining ListenerIds and even wrappers generate their own ListenerIds but combine them with the ones they get returned from inner transports

Why not just the current solution that replaces the TypeId? I think two wrappers of the same type, but with different inner transports (e.g. `DnsConfig<A>` and `DnsConfig<B>`) will have different TypeIds so that should be enough, no?

Reply:
> Yes we clone "leaf" transports like tcp. For example in `development_transport`, we create two tcp transports wrapped in a dns transport, and then wrap one in a WsConfig and combine them in a `OrTransport`. If we would only bubble up the ids generated by the tcp transport, both "or_transport"-ed transports would generate the same Ids:

Looking at the implementation of `DnsConfig`, the inner transport is `Arc<Mutex>`'d, meaning there is only one inner tcp transport, regardless how many times the wrapped DNS transport is cloned. Consequently, there is only one source of `ListenerId`s and there should never be a collision.

> > An alternative solution would be to offer a way of combining ListenerIds and even wrappers generate their own ListenerIds but combine them with the ones they get returned from inner transports
> 
> Why not just the current solution that replaces the TypeId? I think two wrappers of the same type, but with different inner transports (e.g. `DnsConfig<A>` and `DnsConfig<B>`) will have different TypeIds so that should be enough, no?

Yes, it will be enough but I think it would be better if we can find a solution that doesn't require knowledge of the implementation details of `ListenerId` at various places, i.e. when wrapping a `Transport`. Like you said elsewhere, it is not enforced through the type system so chances of it being forgotten are definitely non-zero.

Taking a step back: With transports getting a `poll` function, cloning them should certainly be the exception and if you want to clone one, you will need to `Arc<Mutex>` it. From this perspective, every transport that actually creates listeners (what I called leaf transports) should only ever exist once.

Reply:
I have moved the discussion about Cloning transports to a separate thread https://github.com/libp2p/rust-libp2p/pull/2652/files#r884914910. This way we can keep this thread about the ListenerId API and come back to it once we agree on whether we can / have to construct multiple transports of the same type or not.

Reply:
Assuming we can't clone them, then I'd be in favor of exploring the idea of randomly generate `ListenerId`s because it will result in a smaller API.

Reply:
Done in a357d71.